oauth2/oauth2-jose/src/main/java/org/springframework/security/oauth2/jwt/NimbusJwtDecoder.java

```diff
@@ -283,16 +280,62 @@ public JwkSetUriJwtDecoderBuilder cache(Cache cache) {
 		}
 
 		JWSKeySelector<SecurityContext> jwsKeySelector(JWKSource<SecurityContext> jwkSource) {
-			if (this.signatureAlgorithms.isEmpty()) {
-				return new JWSVerificationKeySelector<>(JWSAlgorithm.RS256, jwkSource);
+			Set<SignatureAlgorithm> algorithms = new HashSet<>();
+			if (!this.signatureAlgorithms.isEmpty()) {
+				algorithms.addAll(this.signatureAlgorithms);
 			} else {
-				Set<JWSAlgorithm> jwsAlgorithms = new HashSet<>();
-				for (SignatureAlgorithm signatureAlgorithm : this.signatureAlgorithms) {
-					JWSAlgorithm jwsAlgorithm = JWSAlgorithm.parse(signatureAlgorithm.getName());
-					jwsAlgorithms.add(jwsAlgorithm);
+				algorithms.addAll(fetchSignatureAlgorithms());
+			}
+
+			if (algorithms.isEmpty()) {
+				algorithms.add(SignatureAlgorithm.RS256);
+			}
+
+			Set<JWSAlgorithm> jwsAlgorithms = new HashSet<>();
+			for (SignatureAlgorithm signatureAlgorithm : algorithms) {
+				jwsAlgorithms.add(JWSAlgorithm.parse(signatureAlgorithm.getName()));
+			}
+
+			return new JWSVerificationKeySelector<>(jwsAlgorithms, jwkSource);
+		}
+
+		private Set<SignatureAlgorithm> fetchSignatureAlgorithms() {
+			if (StringUtils.isEmpty(jwkSetUri)) {
+				return Collections.emptySet();
+			}
+			try {
+				return parseAlgorithms(JWKSet.load(toURL(jwkSetUri), 5000, 5000, 0));
+			} catch (Exception ex) {
+				log.error("Failed to load Signature Algorithms from remote JWK source.");
+				return Collections.emptySet();
```

Comment:
How about creating a new exception class called "JwkException" for JWK related problems? I could use a similar message and throw that particular exception. Nimbus already provides a JwkException, but it is a checked exception. I would like to include this one within Spring Security as an unchecked exception.

Reply:
If we create exceptions, it's important that it be clear whether it's the client's problem or the server's problem. That's one thing that's nice about `IllegalArgumentException` vs `IllegalStateException`.

Before we add a type, though, what's the use case you are trying to address? For example, is there a need to catch this exception and handle it differently from other runtime exceptions?

Reply:
There is no use case I can think of where this would be caught and handled differently. Would you recommend to use IllegalStateException over IllegalArgumentException? I think IllegalStateException would be more appropriate.

Reply:
I typically use `IllegalArgumentException` to indicate user error, e.g. the application ought to be configuring itself differently.